docs/architecture.md

```diff
@@ -0,0 +1,223 @@
+
+### Encapsulation?
+
+Encapsulation is not needed, since it is 100% certain which components a system will read from and write to. Systems are very much analagous to real-world laws of the physics. The real world won't tell a falling apple to update; it's not the apple's responsibility to move itself. The world doesn't care that the apple is even an apple; gravity will affect any object with mass regardless of what it is.
+
+### Updating the world
+
+Most mainstream game engines update the world per-object like this:
+
+```
+void world_update(world) {
+    for (each object in world) {
+        object.update()
+    }
+}
+
+```
+
+An ECS engine will more resemble this:
+
+```
+void world_update(world) {
+    gravity_system(world);
+    move_system(world);
+}
+
+void gravity_system(world) {
+    for (each entity with mass and velocity) {
+        // accelerate the masses together
+    }
+}
+
+void move_system(world) {
+    for (each entity with velocity and position) {
+        position += velocity
+    }
+}
+
+...
+```
+
+Systems are called in a specific order to update the entire world to the next frame. Some systems may depend on each other, and other systems can run in parallel.
+
+The EnTT library provides many features making iteration convenient.
+
+### Responsibility of Systems
+
+Ideally, components are raw data that don't describe any behaviour. However, it's acceptable for them to have constructors, destructors, member functions, and RAII members as long as components are only modified by very specific systems. This includes creating and destroying components. Many systems reading the same component is okay.
+
+EnTT allows polymorphically accessing components, to allow for deleting, copying, and listing all components an entity has. These features should never be used outside of debugging, since they potentially modify components and calls functions outside of their respective systems.
+
+In short, **components must only be modified by a few select systems**. There's a few other practical reasons why:
+
+* Components may need to be synced with external references: pointers, GPU resourses, or physics engine objects. Some of these require specific threads.
+* Multiple Systems that modify different components can run in parallel without mutexes.
+
+There's a few important operations that need to be addressed however:
+
+* **Deleting entire entities**: Deleting an entity means modifying every component container the entity is associated with. Entities can be marked for deletion with a component, or external queue. 'Deleter' Systems can then delete their respective components near the end of the frame.
+* **Copying entire entities**: It may seem convenient to iterate all of an entity's components and copy them into another entity, but this breaks the rules. Copies of an entity should be created by the same system as the original entity. A dedicated 'mark for copy' and copy systems may be practical depending on the case.
```

Comment:
Why's this?

Reply:
These are sort of on-going issues in ECS discussions. The solutions here is what's used in the project, and is a more 'pure' ECS approach.